### PR TITLE
fix(coarse_tile): clear stale Inductor IR caches after _divide_range mutations

### PR DESCRIPTION
**Pull request by dgrove-oss**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md
).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

`_divide_ranges` mutates the `ranges` field on a frozen `Loops`/`Reduction`
dataclass and updates `size`/`stride` on the associated `Layout`, but several
`cache_on_self` / `cache_on_self_and_args` caches on those objects memoize
values derived from those fields.  Without invalidation the downstream
scheduler reads stale sizes and produces incorrect work-division plans.

The fix adds explicit invalidation after each mutation:

| Object | Caches cleared |
|---|---|
| `Loops` | `get_free_symbol_uses`, `inner_fn_str`, `inner_fn_opcount` |
| `Reduction` | `get_free_symbol_uses` (separate key from `Loops`) |
| `Layout` | `get_free_symbol_uses` |
| `ComputedBuffer` | `get_free_symbol_uses`, `get_default_sizes_body` |

Because `Loops`, `Reduction`, and `Layout` are frozen dataclasses,
`cache_on_self` stores results via `object.__setattr__` to bypass the freeze
guard.  Clearing must correspondingly use `object.__delattr__` — plain
`delattr()` raises `FrozenInstanceError`.

To avoid hardcoding private cache-key strings that would silently no-op if
upstream ever renames a method, the keys are derived at module load time by
inspecting the `key` free variable in each cached method's `.clear_cache()`
closure.  This turns any upstream rename into a loud `AttributeError` on
import rather than a silent correctness regression.

#### Which issue(s) this PR is related to:

Part of #1358 

#### Special notes for your reviewer:

`cache_on_self` / `cache_on_self_and_args` both expose a `.clear_cache(instance)`
helper, but it uses plain `delattr()` internally and therefore cannot be called
directly on frozen-dataclass instances.  The `_cache_key()` helper introduced
here extracts the key string from that closure so we can use `object.__delattr__`
ourselves, while still binding to the method name symbolically rather than
copy-pasting a string literal.

#### Does this PR introduce a user-facing change?

No. This is an internal IR-pass correctness fix; there is no change to the
public API or operator support.

#### Additional note:

`Layout` was refactored upstream (custom `__init__` + property setters replacing
declared dataclass fields) between the original branch base and the rebase onto
current `main`.  The frozen-dataclass invariant still holds for `Layout` — its
`__delattr__` is the dataclass-generated one — but it happens to allow
`object.__delattr__` on dynamically-added cache attributes because it has zero
declared `__dataclass_fields__` to protect.  The `_clear_cache` helper is
correct in either case.

### CODE REVIEW

**Comment by marnold-ibm** [comment]
The code looks good.

Claude is suggesting a test could be added.


## Overall

This is a correct, well-engineered fix. The cache invalidation is complete (all
seven cached methods that read `ranges`, `size`, or `stride` are covered), the
frozen-dataclass bypass is correct, and the closure-introspection approach for
deriving cache keys at import time is strictly better than hardcoding strings —
any upstream rename becomes a loud `AttributeError` on import rather than a
silent correctness regression. All CI checks pass. Approve.

## Test suggestion

The existing `TestDivideRanges` tests in `tests/inductor/test_coarse_tiling.py`
verify that `_divide_ranges` produces the right mutation, but none of them
exercise the stale-cache scenario: prime a cache before the call, then verify
the cached value is gone after.

The reason is that `_make_pointwise` and `_make_op` use `MagicMock`, and mocks
are not frozen dataclasses — the `cache_on_self`/`cache_on_self_and_args`
machinery never activates on them. To test cache invalidation you need real IR
objects.

### How to write the test

Real `Pointwise` and `ComputedBuffer` instances construct cleanly with no
special setup:

```python
import sympy
from sympy import Integer
from torch._inductor.ir import Pointwise, ComputedBuffer, FixedLayout
import torch

N = sympy.Symbol("N", positive=True)

pw = Pointwise(
    device=torch.device("cpu"),
    dtype=torch.float16,
    inner_fn=lambda index: sympy.Integer(1),
    ranges=[N, Integer(32)],
)
layout = FixedLayout(torch.device("cpu"), torch.float16, [N, Integer(32)])
op = ComputedBuffer(name="buf0", layout=layout, data=pw)
```

With a symbolic range `N`, `pw.get_free_symbol_uses()` returns `{N}` before
the divide and `{}` after (since the divided range is a concrete integer).
That difference is what the test should assert.

### Suggested test (add to `TestDivideRanges`)

```python
def test_cache_invalidated_after_divide(self):
    # Use a symbolic range so free-symbol content changes after division.
    N = sympy.Symbol("N", positive=True)
    pw = Pointwise(
        device=torch.device("cpu"),
        dtype=torch.float16,
        inner_fn=lambda index: sympy.Integer(1),
        ranges=[N, Integer(32)],
    )
    layout = FixedLayout(
        torch.device("cpu"), torch.float16, [N, Integer(32)]
    )
    op = ComputedBuffer(name="buf0", layout=layout, data=pw)

    # Prime the cache.
    before = pw.get_free_symbol_uses()
    self.assertIn(N, before)

    # Divide — this should clear the cache.
    _divide_ranges(op, Integer(4), tiled_dims=[0])

    # Cache must be gone; fresh call reflects the new (concrete) ranges.
    after = pw.get_free_symbol_uses()
    self.assertNotIn(N, after)
```

A second test for `Reduction` is worth adding to cover the separate
`_REDUCTION_FREE_SYMS_KEY` path. Note that `Reduction.get_free_symbol_uses`
calls `super()`, so a single call to it populates *both*
`__Loops_get_free_symbol_uses_cache` and `__Reduction_get_free_symbol_uses_cache`
on the instance. The PR clears both keys, and asserting on
`red.get_free_symbol_uses()` after the divide exercises both in one call.

```python
def test_cache_invalidated_after_divide_reduction(self):
    N = sympy.Symbol("N", positive=True)
    from torch._inductor.ir import Reduction
    from torch._inductor.ir import ReductionHint

    red = Reduction(
        device=torch.device("cpu"),
        dtype=torch.float16,
        inner_fn=lambda index, rindex: sympy.Integer(1),
        ranges=[N],
        reduction_ranges=[Integer(128)],
        reduction_type="sum",
        src_dtype=torch.float16,
        reduction_hint=ReductionHint.DEFAULT,
    )
    layout = FixedLayout(
        torch.device("cpu"), torch.float16, [N]
    )
    op = ComputedBuffer(name="buf0", layout=layout, data=red)

    before = red.get_free_symbol_uses()
    self.assertIn(N, before)

    _divide_ranges(op, Integer(4), tiled_dims=[0])

    after = red.get_free_symbol_uses()
    self.assertNotIn(N, after)
```

### Why this matters

Without these tests, the `_clear_cache` calls could be silently dropped (e.g.
during a refactor) and no existing test would catch it — the mock-based tests
would still pass. These tests lock in the invariant that the caches are actually
cleared, not just that the mutation produces the right value.

**Comment by dgrove-oss** [comment]
Test improved in lines with  the reviewer suggestion.

**Review by tardieu — APPROVED** [review]
lgtm

**Review by tardieu — APPROVED** [review]
lgtm
